src/hash_join.rs

```diff
@@ -0,0 +1,474 @@
+//! SQL-like join implementation of two (non-sorted) iterators.
+//!
+//! The hash join strategy requires the right iterator can be collected to a `HashMap`. The left
+//! iterator can be arbitrarily long. It is therefore asymmetric (iterators cannot be swapped), as
+//! distinct from [the merge join strategy](merge_join/index.html), which is symmetric.
+//!
+//! The fact that iterators do not need to be sorted makes it very efficient and particularly
+//! suitable for [star schema](https://en.wikipedia.org/wiki/Star_schema) or [snowflake
+//! schema](https://en.wikipedia.org/wiki/Snowflake_schema) joins.
+//!
+//! The supported join types:
+//!
+//! * [`INNER JOIN`](trait.Itertools.html#method.hash_join_inner) - an intersection between the
+//! left and the right iterator.
+//! * [`LEFT EXCL JOIN`](trait.Itertools.html#method.hash_join_left_excl) - a difference
+//! between the left and the right iterator (not directly in SQL).
+//! * [`LEFT OUTER JOIN`](trait.Itertools.html#method.hash_join_left_outer) - a union of `INNER
+//! JOIN` and `LEFT EXCL JOIN`.
+//! * [`RIGHT EXCL JOIN`](trait.Itertools.html#method.hash_join_right_excl) - a difference
+//! between the right and the left iterator (not directly in SQL).
+//! * [`RIGHT OUTER JOIN`](trait.Itertools.html#method.hash_join_right_outer) - a union of `INNER
+//! JOIN` and `RIGHT EXCL JOIN`.
+//! * [`FULL OUTER JOIN`](trait.Itertools.html#method.hash_join_full_outer) - a union of `INNER
+//! JOIN`, `LEFT EXCL JOIN` and `RIGHT EXCL JOIN`.
+
+use std::collections::hash_map::{HashMap, IntoIter,};
+use std::mem;
+use std::rc::Rc;
+use std::hash::Hash;
+use super::EitherOrBoth::{self, Right, Left, Both};
+
+/// An iterator adaptor that [inner
+/// joins](https://en.wikipedia.org/wiki/Join_%28SQL%29#Inner_join) the two base iterators. The
+/// resulting iterator is the intersection of the two base iterators.
+///
+/// The base iterators do *not* need to be sorted. The right base iterator is loaded into HashMap
+/// and thus must be unique on the join key (e.g. by
+/// [grouping](trait.Itertools.html#method.group_by), if necessary) to produce the correct
+/// results. The left base iterator do not need to be unique on the key.
+///
+/// The left base iterator element type must be `(K, LV)`, where `K: Hash + Eq`. 
+/// The right base iterator element type must be `(K, RV)`, where `K: Hash + Eq`.
+///
+/// Iterator element type is `(LV, std::rc::Rc<RV>)`.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
```

Comment:
To be helpful, stress that the whole  `right` iterator is consumed when this JoinInner value is created.